### PR TITLE
Add test

### PR DESCRIPTION
**Pull request by HirotoShioi**
This PR introduces test suites for the `DEPQ` library.

It uses [QuickCheck](https://en.wikipedia.org/wiki/QuickCheck) library, which enables property-based testing.

### How to run the test

```
$ stack test
```

or

```
$ stack test depq:test:depq-test
```
Tests are split into 4 categories:

### Validity testing

This tests that 
1. `DEPQ` that are randomly generated by QuickCheck are always valid.
2. Any functions that returns `DEPQ` (e.g. `insert`, `fromList`, `deleteMin`) will always return a valid `DEPQ`. 

### Post condition

This test that given function will always return an expected value. For instance, if you insert a key value pair into given `DEPQ`, the size of `DEPQ` is the same or larger than the original dichotomy tree.

### Metamorphic testing

This test that multiple set of functions will always return an expected value. For instance, what happens when we insert two random key value pairs into `DEPQ`?

### Model based testing

Here, we test against the model (i.e. Something that would behave similar to DEPQ and is well tested).
The model of `DEPQ` is `Map Int (p, a)`. We also introduce set of functions which would behave the same way that given `DEPQ` should.

### CODE REVIEW

File: src/Data/DEPQ.hs [hunk 1]
```diff
@@ -45,6 +48,13 @@ data DEPQ p a = DEPQ {
 instance (NFData p, NFData a) => NFData (DEPQ p a) where
   rnf (DEPQ mi ma) = rnf mi `seq` rnf ma
 
+instance (Ord p, Arbitrary p, Arbitrary a) => Arbitrary (DEPQ p a) where
+  arbitrary = fromList <$> (arbitrary :: Gen [(Int, p, a)])
+  -- Convert given DEPQ into list, shrink it, then convert it back
+  shrink depq = map fromList $ shrink $ toList depq
+   where
+     toList :: DEPQ p a -> [(Int, p, a)]
+     toList (DEPQ p _) = P.toList p
```
Comment (HirotoShioi) [inline src/Data/DEPQ.hs:57]: Here, we define Arbitrary instance of `DEPQ p a`.

File: src/Data/DEPQ.hs [hunk 2]
```diff
@@ -79,6 +88,9 @@ fromList = foldl insf empty where
 null :: DEPQ p v -> Bool
 null (DEPQ mi ma) = P.null mi && P.null ma
 
+-- | Is the DEPQ valid ?
+valid :: (Ord p) => DEPQ p v -> Bool
+valid (DEPQ mi ma) = P.valid mi && P.valid ma
```
Comment (HirotoShioi) [inline src/Data/DEPQ.hs:93]: Exported from `Data.IntPSQ`. We use this function for validity testing.

**Review by ocramz — APPROVED** [review]
Very nice, thank you so much for this!

**Comment by ocramz** [comment]
Closes #1

File: test/Spec.hs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+      | k1 > k2 = D.findMin depq === Just t2
+      | otherwise = D.findMin depq === Just t1
+  where
+    depq = D.insert k2 p2 v2 $ D.insert k1 p1 v1 D.empty
+
+--------------------------------------------------------------------------------
+-- Model based testing
+-- Here, we test against a data structure that would act as an model
+-- (i.e. Something that would behave similar to DEPQ and is well tested)
+--------------------------------------------------------------------------------
```
Comment (HirotoShioi) [inline test/Spec.hs:198]: Model based testing might be a bit of over-engineering. 

1. It's hard to understand the concept of it.
2. It's re-implementing set of functions.
3. Maintaining them might become difficult.

If you are willing to add test on other categories (e.g. post-condition, metamorphic) you might as well drop this one.

File: depq.cabal [hunk 4]
```diff
@@ -30,19 +30,24 @@ library
...
+  hs-source-dirs:
+      test
+  ghc-options: -threaded -rtsopts -with-rtsopts=-N -Wall
+  build-depends:
+      base >=4.7 && <5
+    , depq
+    , QuickCheck
+    , hspec
+    , containers
+  default-language: Haskell2010
```
Comment (HirotoShioi) [inline depq.cabal:53]: This enables testing.

**Review by HirotoShioi — COMMENTED** [review]
Added some comments for clarity

**Pull request merged**